From 2bb43435486dfae3126a4aa5696aec1da216c2c8 Mon Sep 17 00:00:00 2001
From: Eric Mertens <emertens@gmail.com>
Date: Thu, 28 Sep 2023 11:33:45 -0700
Subject: [PATCH 1/3] Add support for 337 RPL_WHOISTEXT

This message is in WHOIS replies for users with +I
hide_idle_time set.
---
 upstream.go  | 2 +-
 xirc/xirc.go | 1 +
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/upstream.go b/upstream.go
index 9173c3a..00f72cf 100644
--- a/upstream.go
+++ b/upstream.go
@@ -1587,7 +1587,7 @@ func (uc *upstreamConn) handleMessage(ctx context.Context, msg *irc.Message) err
 		}
 
 		dc.SendMessage(ctx, msg)
-	case xirc.RPL_WHOISCERTFP, xirc.RPL_WHOISREGNICK, irc.RPL_WHOISUSER, irc.RPL_WHOISSERVER, irc.RPL_WHOISCHANNELS, irc.RPL_WHOISOPERATOR, irc.RPL_WHOISIDLE, xirc.RPL_WHOISSPECIAL, xirc.RPL_WHOISACCOUNT, xirc.RPL_WHOISACTUALLY, xirc.RPL_WHOISHOST, xirc.RPL_WHOISMODES, xirc.RPL_WHOISSECURE:
+	case xirc.RPL_WHOISCERTFP, xirc.RPL_WHOISREGNICK, irc.RPL_WHOISUSER, irc.RPL_WHOISSERVER, irc.RPL_WHOISCHANNELS, irc.RPL_WHOISOPERATOR, irc.RPL_WHOISIDLE, xirc.RPL_WHOISSPECIAL, xirc.RPL_WHOISACCOUNT, xirc.RPL_WHOISTEXT, xirc.RPL_WHOISACTUALLY, xirc.RPL_WHOISHOST, xirc.RPL_WHOISMODES, xirc.RPL_WHOISSECURE:
 		dc, cmd := uc.currentPendingCommand("WHOIS")
 		if cmd == nil {
 			return fmt.Errorf("unexpected WHOIS reply %q: no matching pending WHOIS", msg.Command)
diff --git a/xirc/xirc.go b/xirc/xirc.go
index 7c809e3..9a1c02c 100644
--- a/xirc/xirc.go
+++ b/xirc/xirc.go
@@ -26,6 +26,7 @@ const (
 	RPL_CREATIONTIME  = "329"
 	RPL_WHOISACCOUNT  = "330"
 	RPL_TOPICWHOTIME  = "333"
+	RPL_WHOISTEXT     = "337"
 	RPL_WHOISACTUALLY = "338"
 	RPL_WHOSPCRPL     = "354"
 	RPL_WHOISHOST     = "378"

From e356274f2c97f56d2708860d6b1c76df9aaa7ccc Mon Sep 17 00:00:00 2001
From: Eric Mertens <emertens@gmail.com>
Date: Sun, 13 Aug 2023 09:04:59 -0700
Subject: [PATCH 2/3] Add support for solanum-specific caps

---
 downstream.go | 12 ++++++++++++
 upstream.go   |  4 ++++
 2 files changed, 16 insertions(+)

diff --git a/downstream.go b/downstream.go
index f48653e..af15e75 100644
--- a/downstream.go
+++ b/downstream.go
@@ -252,6 +252,10 @@ var needAllDownstreamCaps = map[string]string{
 	"multi-prefix":     "",
 
 	"draft/extended-monitor": "",
+
+	"solanum.chat/identify-msg": "",
+	"solanum.chat/oper":         "",
+	"solanum.chat/realhost":     "",
 }
 
 // passthroughIsupport is the set of ISUPPORT tokens that are directly passed
@@ -492,6 +496,14 @@ func (dc *downstreamConn) SendMessage(ctx context.Context, msg *irc.Message) {
 				supported = dc.caps.IsEnabled("account-tag")
 			case "batch":
 				supported = dc.caps.IsEnabled("batch")
+			case "solanum.chat/identified":
+				supported = dc.caps.IsEnabled("solanum.chat/identify-msg")
+			case "solanum.chat/oper":
+				supported = dc.caps.IsEnabled("solanum.chat/oper")
+			case "solanum.chat/realhost":
+				supported = dc.caps.IsEnabled("solanum.chat/realhost")
+			case "solanum.chat/ip":
+				supported = dc.caps.IsEnabled("solanum.chat/realhost")
 			}
 			if !supported {
 				delete(msg.Tags, name)
diff --git a/upstream.go b/upstream.go
index 00f72cf..3765083 100644
--- a/upstream.go
+++ b/upstream.go
@@ -45,6 +45,10 @@ var permanentUpstreamCaps = map[string]bool{
 
 	"draft/account-registration": true,
 	"draft/extended-monitor":     true,
+
+	"solanum.chat/oper":         true,
+	"solanum.chat/realhost":     true,
+	"solanum.chat/identify-msg": true,
 }
 
 // storableMessageTags is the static list of message tags that will cause

From 08a6c0b91beaf40521676305531285de1fad6c23 Mon Sep 17 00:00:00 2001
From: Eric Mertens <emertens@gmail.com>
Date: Thu, 16 May 2024 20:12:53 -0700
Subject: [PATCH 3/3] fix MODE and WHOWAS

---
 downstream.go | 82 +++++----------------------------------------------
 upstream.go   | 22 +++++++++++++-
 2 files changed, 29 insertions(+), 75 deletions(-)

diff --git a/downstream.go b/downstream.go
index af15e75..24585f1 100644
--- a/downstream.go
+++ b/downstream.go
@@ -2003,85 +2003,12 @@ func (dc *downstreamConn) handleMessageRegistered(ctx context.Context, msg *irc.
 
 		uc.SendMessageLabeled(ctx, dc.id, msg)
 	case "MODE":
-		var name string
-		if err := parseMessageParams(msg, &name); err != nil {
-			return err
-		}
-
-		var modeStr string
-		if len(msg.Params) > 1 {
-			modeStr = msg.Params[1]
-		}
-
-		if dc.casemap(name) == dc.nickCM {
-			if modeStr != "" {
-				uc, err := dc.upstreamForCommand(msg.Command)
-				if err != nil {
-					return err
-				}
-				uc.SendMessageLabeled(ctx, dc.id, msg)
-			} else {
-				var userMode string
-				if uc := dc.upstream(); uc != nil {
-					userMode = string(uc.modes)
-				}
-
-				dc.SendMessage(ctx, &irc.Message{
-					Command: irc.RPL_UMODEIS,
-					Params:  []string{dc.nick, "+" + userMode},
-				})
-			}
-			return nil
-		}
-
 		uc, err := dc.upstreamForCommand(msg.Command)
 		if err != nil {
 			return err
 		}
 
-		if !uc.isChannel(name) {
-			return ircError{&irc.Message{
-				Command: irc.ERR_USERSDONTMATCH,
-				Params:  []string{dc.nick, "Cannot change mode for other users"},
-			}}
-		}
-
-		if modeStr != "" {
-			params := []string{name, modeStr}
-			params = append(params, msg.Params[2:]...)
-			uc.SendMessageLabeled(ctx, dc.id, &irc.Message{
-				Command: "MODE",
-				Params:  params,
-			})
-		} else {
-			ch := uc.channels.Get(name)
-			if ch == nil {
-				// we're not on that channel, pass command to upstream
-				uc.SendMessageLabeled(ctx, dc.id, msg)
-				return nil
-			}
-
-			if ch.modes == nil {
-				// we haven't received the initial RPL_CHANNELMODEIS yet
-				// ignore the request, we will broadcast the modes later when we receive RPL_CHANNELMODEIS
-				return nil
-			}
-
-			modeStr, modeParams := ch.modes.Format()
-			params := []string{dc.nick, name, modeStr}
-			params = append(params, modeParams...)
-
-			dc.SendMessage(ctx, &irc.Message{
-				Command: irc.RPL_CHANNELMODEIS,
-				Params:  params,
-			})
-			if ch.creationTime != "" {
-				dc.SendMessage(ctx, &irc.Message{
-					Command: xirc.RPL_CREATIONTIME,
-					Params:  []string{dc.nick, name, ch.creationTime},
-				})
-			}
-		}
+		uc.SendMessageLabeled(ctx, dc.id, msg)
 	case "TOPIC":
 		var name string
 		if err := parseMessageParams(msg, &name); err != nil {
@@ -2330,6 +2257,13 @@ func (dc *downstreamConn) handleMessageRegistered(ctx context.Context, msg *irc.
 			return err
 		}
 
+		uc.enqueueCommand(dc, msg)
+	case "WHOWAS":
+		uc, err := dc.upstreamForCommand(msg.Command)
+		if err != nil {
+			return err
+		}
+
 		uc.enqueueCommand(dc, msg)
 	case "PRIVMSG", "NOTICE", "TAGMSG":
 		var targetsStr, text string
diff --git a/upstream.go b/upstream.go
index 3765083..79528a5 100644
--- a/upstream.go
+++ b/upstream.go
@@ -483,6 +483,12 @@ func (uc *upstreamConn) abortPendingCommands() {
 					Command: irc.RPL_ENDOFWHOIS,
 					Params:  []string{dc.nick, nick, "Command aborted"},
 				})
+			case "WHOWAS":
+				nick := pendingCmd.msg.Params[0]
+				dc.SendMessage(ctx, &irc.Message{
+					Command: irc.RPL_ENDOFWHOWAS,
+					Params:  []string{dc.nick, nick, "Command aborted"},
+				})
 			case "AUTHENTICATE":
 				dc.endSASL(ctx, &irc.Message{
 					Command: irc.ERR_SASLABORTED,
@@ -513,7 +519,7 @@ func (uc *upstreamConn) sendNextPendingCommand(cmd string) {
 
 func (uc *upstreamConn) enqueueCommand(dc *downstreamConn, msg *irc.Message) {
 	switch msg.Command {
-	case "LIST", "WHO", "WHOIS", "AUTHENTICATE", "REGISTER", "VERIFY":
+	case "LIST", "WHO", "WHOIS", "WHOWAS", "AUTHENTICATE", "REGISTER", "VERIFY":
 		// Supported
 	default:
 		panic(fmt.Errorf("Unsupported pending command %q", msg.Command))
@@ -1593,6 +1599,11 @@ func (uc *upstreamConn) handleMessage(ctx context.Context, msg *irc.Message) err
 		dc.SendMessage(ctx, msg)
 	case xirc.RPL_WHOISCERTFP, xirc.RPL_WHOISREGNICK, irc.RPL_WHOISUSER, irc.RPL_WHOISSERVER, irc.RPL_WHOISCHANNELS, irc.RPL_WHOISOPERATOR, irc.RPL_WHOISIDLE, xirc.RPL_WHOISSPECIAL, xirc.RPL_WHOISACCOUNT, xirc.RPL_WHOISTEXT, xirc.RPL_WHOISACTUALLY, xirc.RPL_WHOISHOST, xirc.RPL_WHOISMODES, xirc.RPL_WHOISSECURE:
 		dc, cmd := uc.currentPendingCommand("WHOIS")
+
+		if dc == nil || cmd == nil {
+		        dc, cmd = uc.currentPendingCommand("WHOWAS")
+		}
+
 		if cmd == nil {
 			return fmt.Errorf("unexpected WHOIS reply %q: no matching pending WHOIS", msg.Command)
 		} else if dc == nil {
@@ -1608,6 +1619,15 @@ func (uc *upstreamConn) handleMessage(ctx context.Context, msg *irc.Message) err
 			return nil
 		}
 
+		dc.SendMessage(ctx, msg)
+	case irc.RPL_ENDOFWHOWAS:
+		dc, cmd := uc.dequeueCommand("WHOWAS")
+		if cmd == nil {
+			return fmt.Errorf("unexpected RPL_ENDOFWHOWAS: no matching pending WHOWAS")
+		} else if dc == nil {
+			return nil
+		}
+
 		dc.SendMessage(ctx, msg)
 	case "INVITE":
 		var nick, channel string
